HWP script addition

yukisakurai opened the pull request:
Adding the HWP angle calculator script to sotodlib.
The update_hwp_angle.py script follows pipeline products of "HWP angle" in the First light pipeline document. 
The HWP angle g3 (Level 3) file is generated by reading the L2 SO HK g3 file.
We confirmed it works as expected for the SAT1 and TSAT lab. data.
Adding also HWPSS pipeline products and simple demod. function.
Tutorial notebook is here: https://github.com/yukisakurai/so_pwg_tutorial/blob/main/tutorial_hwp.ipynb

kmharrington commented on sotodlib/hwp/hwp.py:
```diff
@@ -0,0 +1,146 @@
+    filt = tod_ops.filters.high_pass_sine2(cutoff=0.1)
+    signal_prefilt = np.array(tod_ops.fourier_filter(aman, filt, detrend=None, signal_name=signal))
+    
+    bins=361
```
this should probably be a kwarg instead of hardcoded.

yukisakurai replied:
Agreed. Done.
https://github.com/simonsobs/sotodlib/blob/d87f251fbff1a74ef601c17c79312bf1610f7562/sotodlib/hwp/hwp.py#L21

kmharrington commented on sotodlib/hwp/hwp.py:
```diff
@@ -0,0 +1,146 @@
+    elif mode == 3: func = func_hwpss_all
+    for i in range(aman.dets.count):
+        popt, pcov = curve_fit(func, hwp_angle_bins, hwpss[i], maxfev=1000000)
+        aman.ext_hwpss[i] += func(aman.hwp_angle, *popt)
```
I think you want this to be an `=` instead of a `+=`? Technically `wrap_new` should fill with zeros but this is something you're just calculating right?

yukisakurai replied:
Thanks, fixed
https://github.com/simonsobs/sotodlib/blob/80b784143dd5fb3b49f965c6ae90ed83a1962b91/sotodlib/hwp/hwp.py#L52

kmharrington commented on sotodlib/hwp/update_hwp_angle.py:
```diff
@@ -0,0 +1,517 @@
+        """
+        self._start = start
+        self._end = end
+        hwp_keys=[
```
These need to be in a config file somewhere. `load_range` takes config files as well as fields/aliases. We don't want to have to use different versions of sotodlib if someone changes a feed name. for this function and `load_file` as well.

yukisakurai replied:
Implemented to work with config yaml file. I put my config file here: `simons1:/mnt/so1/users/ysakurai/hwp_configs/hwp_config.yaml`
But I made it also work without a config file for convenience.

kmharrington commented on sotodlib/hwp/update_hwp_angle.py:
```diff
@@ -0,0 +1,517 @@
+        """
+
+        
+        self._debug = debug
```
instead of a setting like this we want to use a logger from the python package `logging`. You can look at `sotodlib/__init__.py` or `sotodlib/io/load_smurf.py` for examples. Instead you can use the debug flag to set the level of the logger.

yukisakurai replied:
Implemented logging.

kmharrington commented on sotodlib/hwp/update_hwp_angle.py:
```diff
@@ -0,0 +1,517 @@
+        self._end = 0
+        # Size of pakcets sent from the BBB
+        self._pkt_size = 120 # maybe 120 in the latest version, 150 in the previous version
+        # Number of encoder slits per HWP revolution
```
these seem like a hardware config things. Could we add them to a HWP config file so we can be prepared for it to change with time/device (even if we have no plans of changing things, SO lasts a long time and will have many HWPs). 

This config could include the fields names I comment on below.

yukisakurai replied:
Same as above. Implemented config.

kmharrington reviewed (changes requested):
I wasn't tagged in the review but doing one anyway :-)

Great to see this starting out, I have some higher level comments and them some in-line ones. 

* There are a lot of hard coded numbers / field names that look like they're associated with hardware/hardware settings. These should all go into some sort of configuration file.
* I think the HWP angle calculation math and things can go in `sotodlib/hwp/` but I would move the script section into `sotodlib/site_pipeline/` folder with the rest of them. This is similar to how things work with the make-uncal-beam-map code.

A couple code/style comments: 

* I would remove all the print statements and switch to using a logger made with the `logging` package. That gives the user a lot more control over printing and/or saving logs as this script runs. 
* We try and follow [PEP8 style](https://peps.python.org/pep-0008/) in our code. The ones I'm seeing to change would be long line lengths (we want to wrap at 80 characters) and removing all the single line if statements.

yukisakurai commented on sotodlib/hwp/hwp.py:
```diff
@@ -0,0 +1,146 @@
+    - 2f and 4f amplitudes and phases are extracted by HWPSS fitting of MIQ and MIU.
+    - Amplitudes and phases are fitted by polynominal functions w.r.t each incident angle.
+    Args ---
+    band: MF1=0, MF2=1
```
I agreed and changed it to bandpass.
https://github.com/simonsobs/sotodlib/blob/hwp/sotodlib/hwp/hwp_sim.py#:~:text=if%20bandpass%20%3D%3D%20%22f090%22%3A

yukisakurai commented on sotodlib/hwp/hwp.py:
```diff
@@ -0,0 +1,146 @@
+import os
```
I splited to hwp.py and hwp_sim.py
https://github.com/simonsobs/sotodlib/tree/hwp/sotodlib/hwp/

yukisakurai commented:
> I wasn't tagged in the review but doing one anyway :-)

I added Katie as reviewer. Thank you so much for your comments. 

> Great to see this starting out, I have some higher level comments and them some in-line ones.
> * There are a lot of hard coded numbers / field names that look like they're associated with hardware/hardware settings. These should all go into some sort of configuration file.

I implemented hwp config file

> * I think the HWP angle calculation math and things can go in `sotodlib/hwp/` but I would move the script section into `sotodlib/site_pipeline/` folder with the rest of them. This is similar to how things work with the make-uncal-beam-map code.

I moved to script part (FLP element: update_hwp_angle.py) to `sotodlib/site_pipeline/`
Now tools (FLP products) are stored in `sotodlib/hwp`

> A couple code/style comments:
>
> * I would remove all the print statements and switch to using a logger made with the `logging` package. That gives the user a lot more control over printing and/or saving logs as this script runs.

I Implemented logging 

> * We try and follow [PEP8 style](https://peps.python.org/pep-0008/) in our code. The ones I'm seeing to change would be long line lengths (we want to wrap at 80 characters) and removing all the single line if statements.

I applied autopep8!

kmharrington commented on sotodlib/hwp/g3thwp.py:
```diff
@@ -0,0 +1,713 @@
+import yaml
+
+logger = logging.getLogger(__name__)
+# logging.basicConfig(level=logging.DEBUG)
```
delete stray comment

yukisakurai replied:
deleted

kmharrington reviewed (approved):
I'm going to go ahead and approve this because I think anything else we want changed could be left to later issues. Thank you for implementing all the requests! The new in-line comments are mostly just requests for more docstrings. 

I think the one additional thing would be adding a section to the site-pipeline documentation for the update-hwp-angle script. We should also eventually add a whole HWP section to the documentation for interacting with these scripts. I think the docstrings are useful enough that it's not necessary for this PR but should be on the to-do list. (If you've never compiled / tested docs, I can point you to info there. simons1 has a place you can put them so you can test that it worked). 

I haven't actually tested the code so if @jseibert575 or @msilvafe  have comments on operation those could be useful (either as future issues or for this PR).

yukisakurai commented on sotodlib/site_pipeline/update_hwp_angle.py:
```diff
@@ -0,0 +1,95 @@
+    parser.add_argument(
+        '-d', '--data-dir', action='store', default=None, type=str,
+        help='input data directory')
+    parser.add_argument(
```
added explanation

yukisakurai commented on sotodlib/hwp/hwp.py:
```diff
@@ -0,0 +1,122 @@
+                                     signal_name='signal_demod_wo_lfilt'), 
+              [(0, 'dets'), (1, 'samps')])
+
+    aman.move('signal_demod_prelfilt', '')
```
fixed

yukisakurai commented on sotodlib/hwp/g3thwp.py:
```diff
@@ -0,0 +1,713 @@
+        config_file: str
+            path to config yaml file
+            yaml file should include below info.
+            -----------------------
```
added explanation

yukisakurai commented on sotodlib/site_pipeline/update_hwp_angle.py:
```diff
@@ -0,0 +1,95 @@
+        help="Configuration File for running update_hwp_angle")
+    parser.add_argument(
+        '-t', '--time', action='store', default=None, type=int,
+        help='time range ex) --time [start] [end]',
```
Add explanation in the help and docs
